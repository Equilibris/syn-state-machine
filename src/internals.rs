mod buffer;

pub use buffer::*;

pub trait ParserCursor {
    type Error;
}

pub struct LocError<'a, Loc>(pub &'a str, pub Loc);

pub trait Spanned {
    type Loc;

    fn span(&self) -> Self::Loc;
}

pub trait Finalizer<Out, With> {
    fn finalize(self, value: With) -> std::ops::ControlFlow<Out, Out>;
}

pub struct BlackHoleFinalizer<T>(pub T);

impl<Out, With> Finalizer<Out, With> for BlackHoleFinalizer<Out> {
    fn finalize(self, _: With) -> std::ops::ControlFlow<Out, Out> {
        std::ops::ControlFlow::Break(self.0)
    }
}

pub trait Parse<C: ParserCursor, With>: Sized {
    type Finalizer: Finalizer<Self, With>;

    fn parse(input: &mut ParseBuffer<C>) -> Result<Self::Finalizer, C::Error>;
}

pub trait Peek<C> {
    fn peek(input: &C) -> Option<usize>;
}
pub trait FixedPeek {
    const SKIP: usize;
}
pub trait PeekError<C: ParserCursor> {
    fn error(input: &C) -> C::Error;
}
pub trait CombineError<Other> {
    fn combine(&mut self, other: Other);
}
