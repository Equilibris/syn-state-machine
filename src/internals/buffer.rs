use proc_macro2::Span;

use crate::{internals::Result, Error};

#[derive(Clone, Copy)]
pub struct ParseBuffer<C> {
    pub cursor: C,
}

impl<C> From<C> for ParseBuffer<C> {
    fn from(cursor: C) -> Self {
        Self { cursor }
    }
}

impl<C: Skip> ParseBuffer<C> {
    pub fn peek<T: Peek<C>>(&mut self) -> bool {
        if let Some(x) = T::peek(&self.cursor) {
            self.cursor.skip(x);
            true
        } else {
            false
        }
    }
    pub fn errored_peek<T: Peek<C> + PeekError<C>>(&mut self) -> Result<()> {
        if let Some(x) = T::peek(&self.cursor) {
            self.cursor.skip(x);
            Ok(())
        } else {
            Err(T::error(&self.cursor))
        }
    }
}
impl<C> ParseBuffer<C> {
    pub fn parse<T: Parse<C>>(&mut self) -> Result<T> {
        T::parse(self)
    }
    pub fn call<T, E>(&mut self, function: impl Fn(&mut Self) -> Result<T>) -> Result<T> {
        function(self)
    }
}

impl<C: Spanned> Spanned for ParseBuffer<C> {
    fn span(&self) -> Span {
        self.cursor.span()
    }
}

pub trait Skip {
    fn eof(&self) -> bool;
    fn skip(&mut self, count: usize);
}

pub trait Spanned {
    fn span(&self) -> Span;
}

pub trait Parse<C>: Sized {
    fn parse(input: &mut ParseBuffer<C>) -> Result<Self>;
}
pub trait Peek<C> {
    fn peek(input: &C) -> Option<usize>;
}
pub trait FixedPeek {
    const SKIP: usize;
}
pub trait PeekError<C> {
    fn error(input: &C) -> Error;
}
