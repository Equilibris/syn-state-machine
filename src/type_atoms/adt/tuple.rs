use crate::internals::*;

macro_rules! tuple_impl {
    () => {
        impl<C> Parse<C> for () {
            fn parse(_: &mut ParseBuffer<C>) -> Result<Self> {
                Ok(())
            }
        }
        impl<C> Peek<C> for () {
            fn peek(_: &C) -> Option<usize> {
                Some(0)
            }
        }

        impl FixedPeek for () {
            const SKIP: usize = 0;
        }
        // Do not implement
        // impl PeekError for () {
        //     fn error<'a>(input: &ParseBuffer<'a>) -> Error {
        //         panic!("Atemted to fail infallable peek")
        //         // Error::new(input.span(), "Should never be reached")
        //     }
        // }
    };
    ($last_gen:ident $($gen:ident)*) => {
        impl<Cursor: Clone, $last_gen: Parse<Cursor>, $($gen: Parse<Cursor>,)*> Parse<Cursor> for ($last_gen, $($gen,)* ) {
            fn parse(input: &mut ParseBuffer<Cursor>) -> Result<Self> {
                let mut temp = input.clone();

                let vs = (temp.parse()?, $(temp.parse::<$gen>()?,)*);

                *input = temp;

                Ok(vs)
            }
        }

        impl<Cursor: Skip + Clone, $last_gen: Peek<Cursor>, $($gen: Peek<Cursor>),*> Peek<Cursor> for ($last_gen, $($gen,)*) {

            fn peek(cursor: &Cursor) -> Option<usize> {
                #[allow(unused_mut)]
                let mut v = $last_gen::peek(cursor)?;
                let mut cursor = cursor.clone();

                cursor.skip(v);

                $({
                    let delta = $gen::peek(&cursor)?;
                    cursor.skip(delta);
                    v += delta;
                })*

                dbg!(v);
                Some(v)
            }
        }

        impl<$last_gen: FixedPeek, $($gen: FixedPeek),*> FixedPeek for ($last_gen, $($gen,)*) {
            const SKIP: usize = $last_gen::SKIP $(+ $gen::SKIP)*;
        }
        // Not perfect. In reality the last generic does not have to impl FixedPeek but this is
        // close enougth
        impl<Cursor: Clone + Skip, $last_gen: PeekError<Cursor> + FixedPeek, $($gen: PeekError<Cursor> + FixedPeek),*> PeekError<Cursor> for ($last_gen, $($gen,)*) {
            fn error(cursor: &Cursor) -> Error {
                #[allow(unused_mut)]
                let mut e = $last_gen::error(&cursor);
                let mut cursor = cursor.clone();
                cursor.skip($last_gen::SKIP);

                $({
                    e.combine($gen::error(&cursor));
                    cursor.skip($gen::SKIP);
                })*

                e
            }
        }

        tuple_impl!($($gen)*);
    };
}

tuple_impl!(Z Y X W V U T S R Q P O N M L K J I H G F E D C B A);
// tuple_impl!(AZ AY AX AW AV AU AT AS AR AQ AP AO AN AM AL AK AJ AI AH AG AF AE AD AC AB AA Z Y X W V U T S R Q P O N M L K J I H G F E D C B A);
#[cfg(test)]
mod tests {
    use crate::*;

    insta_match_test!(it_fails_2_tuple, (Ident, FIdent<"world">) : hello *);
    insta_match_test!(it_fails_2_types, (Ident, Ident) : hello *);
    insta_match_test!(it_matches_2_tuple, (Ident, FIdent<"world">) : hello world);
    insta_match_test!(it_steps_back_for_options, (Option<Ident>, Option<Punct>) : <);
    insta_match_test!(it_only_steps_back_on_fail_for_options, (Option<Ident>, Option<Punct>) : hi);
    insta_match_test!(it_steps_back_for_multi_tuples, (Option<Ident>, Option<Punct>, Option<Ident>, Option<Punct>) : hi <>);
    insta_match_test!(it_sums_tuple_backtracking, (Vec<(Punct, Punct, Ident, Ident)>, Punct) : >>h1 h2>>h3 h4 !);
}
