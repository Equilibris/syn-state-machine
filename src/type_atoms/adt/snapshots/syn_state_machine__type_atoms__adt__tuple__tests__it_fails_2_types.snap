---
source: src/type_atoms/adt/tuple.rs
expression: "(insta_match_test!(+ parse (Ident, Ident) : hello *),\n    insta_match_test!(* peek (Ident, Ident) : hello *))"
---
(
    Err(
        Error(
            "Expected ident",
        ),
    ),
    None,
)
